s3/src/main/scala/tamer/s3/Setup.scala

```diff
@@ -4,41 +4,99 @@ package s3
 import com.sksamuel.avro4s.{Decoder, Encoder, SchemaFor}
 import eu.timepit.refined.types.numeric.PosInt
 import zio.stream.ZTransducer
+import zio.{Queue, UIO, ZIO}
 
-import java.time.Duration
+import java.time.format.DateTimeFormatter
+import java.time.{Duration, Instant}
 import scala.util.hashing.MurmurHash3.stringHash
 
-final case class Setup[R, V <: Product: Encoder: Decoder: SchemaFor](
+final case class Setup[
+    R,
+    K <: Product: Encoder: Decoder: SchemaFor,
+    V <: Product: Encoder: Decoder: SchemaFor,
+    S <: Product: Decoder: Encoder: SchemaFor
+](
     bucketName: String,
     prefix: String,
-    afterwards: LastProcessedInstant,
+    override val defaultState: S,
+    override val stateKey: Int,
     transducer: ZTransducer[R, TamerError, Byte, V],
     parallelism: PosInt,
-    zonedDateTimeFormatter: ZonedDateTimeFormatter,
-    minimumIntervalForBucketFetch: Duration
-) extends _root_.tamer.Setup[S3Object, V, LastProcessedInstant](
-      Serde[S3Object](isKey = true).serializer,
+    minimumIntervalForBucketFetch: Duration,
+    maximumIntervalForBucketFetch: Duration,
+    getNextState: (KeysR, S, Queue[Unit]) => UIO[S],
```

Comment:
The `Queue[Unit]` provides a token whenever the file list is changed, client code is free to ignore this, but usually will want to only act whenever new files are written.